common/src/main/java/com/codeforcommunity/auth/Passwords.java

```diff
@@ -0,0 +1,101 @@
+package com.codeforcommunity.auth;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Arrays;
+import java.util.Random;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+/**
+ * A utility class to hash passwords and check passwords vs hashed values. It uses a combination of hashing and unique
+ * salt. The algorithm used is PBKDF2WithHmacSHA1 which, although not the best for hashing password (vs. bcrypt) is
+ * still considered robust and <a href="https://security.stackexchange.com/a/6415/12614"> recommended by NIST </a>.
+ * The hashed value has 256 bits.
+ */
+public class Passwords {
+
+  private static final Random RANDOM = new SecureRandom();
+  private static final int ITERATIONS = 10000;
+  private static final int KEY_LENGTH = 256;
+
+  /**
+   * Returns a random salt to be used to hash a password.
+   *
+   * @return a 16 bytes random salt
+   */
+  public static byte[] getNextSalt(int length) {
+    byte[] salt = new byte[length];
+    RANDOM.nextBytes(salt);
+    return salt;
+  }
+
+  public static byte[] getNextSalt() {
+    return getNextSalt(50);
+  }
+
+  /**
+   * Returns a salted and hashed password using the provided hash.<br>
+   * Note - side effect: the password is destroyed (the char[] is filled with zeros)
+   *
+   * @param password the password to be hashed
+   * @param salt     a 16 bytes salt, ideally obtained with the getNextSalt method
+   *
+   * @return the hashed password with a pinch of salt
+   */
+  public static byte[] hash(char[] password, byte[] salt) {
+    PBEKeySpec spec = new PBEKeySpec(password, salt, ITERATIONS, KEY_LENGTH);
+    Arrays.fill(password, Character.MIN_VALUE);
+    try {
+      SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
```

Comment:
What is the meaning of the String "PBKD...."?

Reply:
> In cryptography, PBKDF1 and PBKDF2 (Password-Based Key Derivation Function 2) are key derivation functions with a sliding computational cost, used to reduce vulnerabilities to brute force attacks. 

[Secret key derivation](https://en.wikipedia.org/wiki/PBKDF2)

Reply:
put PBKDF2... in variable